File: rust-code/rust-python/ex40/ex40.py
class Song(object):
	"""docstring for Song"""
	def __init__(self, lyrics):
		super(Song, self).__init__()
		self.lyrics = lyrics

	def sing_me_a_song(self):
		for line in self.lyrics:
			print(line)


happy_bday = Song(["Happy birthday to you", "I don't want to get sued","So I'll stop right there"])

bulls_on_parade = Song(["They rally around tha family","With pockets full of shells"])

happy_bday.sing_me_a_song()

bulls_on_parade.sing_me_a_song()

class Song(object):

    def __init__(self, disk):
        self.index = 0
        self.disk = disk
        self.jump()

    def next(self):
        ''' next song. '''
        self.index = (self.index + 1) % len(self.disk)
        self.jump()

    def prev(self):
        ''' prev song. '''
        self.index = (self.index - 1) % len(self.disk)
        self.jump()

    def jump(self):
        ''' jump to the song. '''
        self.lyrics = self.disk[self.index]

    def sing_me_a_song(self):
        for line in self.lyrics:
            print(line)

# construct a disk
song1 =  ["Happy birthday to you",
"I don't want to get sued",
"So I'll stop right there"]

song2 = ["They rally around the family",
"With pockets full of shells"
]

song3 = ["Never mind I find",
"Some one like you"
]

disk = [song1, song2, song3]

mycd = Song(disk)
mycd.sing_me_a_song()

mycd.next()
mycd.sing_me_a_song()

mycd.next()
mycd.sing_me_a_song()

mycd.next()
mycd.sing_me_a_song()

mycd.prev()
mycd.sing_me_a_song()

mycd.prev()
mycd.sing_me_a_song()

mycd.prev()
mycd.sing_me_a_song()

mycd.prev()
mycd.sing_me_a_song()

File: rust-code/rust-python/ex40/ex40.rs
fn main(){
    let happy_bday = Song::init(vec!["Happy birthday to you".into(), "I don't want to get sued".into(),"So I'll stop right there".into()]);
    let bulls_on_parade = Song::init(vec!["They rally around tha family".into(),"With pockets full of shells".into()]);
    happy_bday.sing_me_a_song();
    bulls_on_parade.sing_me_a_song();
    let song1 =  Song::init(vec!["Happy birthday to you".into(),"I don't want to get sued".into(),"So I'll stop right there".into()]);
    let song2 =  Song::init(vec!["They rally around the family".into(),"With pockets full of shells".into()]);
    let song3 =  Song::init(vec!["Never mind I find".into(),"Some one like you".into()]);
    let disk = vec![song1, song2, song3];
    let mut mycd = Disk::init(disk);
    let s1=mycd.jump();
    s1.sing_me_a_song();
    let s2=&mut mycd.next();
    s2.sing_me_a_song(); 
    let s3=&mut mycd.next();
    s3.sing_me_a_song();
    let s1=&mut mycd.prev();
    s1.sing_me_a_song();
    let s2=&mut mycd.prev();
    s2.sing_me_a_song(); 
    
}
#[derive(Clone)]
struct Song{
    lyrics:Vec<String>,
}
struct Disk{
    disk:Vec<Song>,
    index:u32,
}
impl Song{
    fn init(lyrics:Vec<String>)->Self{
        Song{lyrics}
    }
    fn sing_me_a_song(&self){
        for line in &self.lyrics{
            println!("{}",line);
        }
    }
}
impl Disk{
    fn init(disk:Vec<Song>)->Self{
        Disk{index:0,disk}
    }
    fn jump(&self)-> Song{
        self.disk[self.index as usize].clone()
    }
    fn next(& mut self)-> Song{
        self.index = (self.index + 1) % self.disk.len() as u32;
        self.jump()
    }
    fn prev(& mut self)->Song{
        self.index = (self.index - 1) % self.disk.len() as u32;
        self.jump()
    }
}